File: android/src/main/java/kashier/wrapper/KashierPaymentResponseParser.java
package kashier.wrapper;

import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;

import io.kashier.sdk.Core.Helpers.ResponseHelper.ResponseHelper;
import io.kashier.sdk.Core.model.Response.I_GenericResponse.I_GenericResponse;
import io.kashier.sdk.Core.model.Response.I_GenericResponse.I_Messages;
import io.kashier.sdk.Core.model.Response.Payment.Card;
import io.kashier.sdk.Core.model.Response.Payment.CardInfo;
import io.kashier.sdk.Core.model.Response.Payment.DSecure;
import io.kashier.sdk.Core.model.Response.Payment.Merchant;
import io.kashier.sdk.Core.model.Response.Payment.Messages;
import io.kashier.sdk.Core.model.Response.Payment.PaymentResponse;
import retrofit2.Response;

public class KashierPaymentResponseParser {
    static private WritableMap getMessagesFromResponse(I_GenericResponse paymentResponse) {
        WritableMap _messagesMap = new WritableNativeMap();
        I_Messages _messages = null;
        if (paymentResponse != null) {
            if (paymentResponse.getMessages() != null) {
                _messages = paymentResponse.getMessages();
            }
            if (paymentResponse.getResponse() != null) {
                if (paymentResponse.getResponse().getMessages() != null) {
                    _messages = paymentResponse.getResponse().getMessages();
                }
                if (paymentResponse.getResponse().getMessage() != null
                        && paymentResponse.getResponse().getMessage().getMessages() != null) {
                    _messages = paymentResponse.getResponse().getMessage().getMessages();
                }
            }
        }
        if (_messages != null) {
            if (_messages.getAr() != null) {
                _messagesMap.putString("ar", _messages.getAr());
            }
            if (_messages.getEn() != null) {
                _messagesMap.putString("en", _messages.getEn());
            }
        }
        return _messagesMap;
    }

    static WritableMap parsePaymentResponse(Response<PaymentResponse> paymentResponse) {
        WritableMap _paymentResponseMap = new WritableNativeMap();
        if (paymentResponse.body() != null) {
            if (ResponseHelper.getResponseStatus(paymentResponse.body()) != null) {
                _paymentResponseMap.putString("status", ResponseHelper.getResponseStatus(paymentResponse.body()).toString());
            }
            _paymentResponseMap.putMap("messages", getMessagesFromResponse(paymentResponse.body()));

            if (paymentResponse.body().getResponse() != null) {
                io.kashier.sdk.Core.model.Response.Payment.Response _response = paymentResponse.body().getResponse();
                WritableMap _payment_response = new WritableNativeMap();
                if (_response.getMethod() != null) {
                    _payment_response.putString("method", _response.getMethod());
                }
                if (_response.getOrderId() != null) {
                    _payment_response.putString("orderId", _response.getOrderId());
                }
                if (_response.getApiKeyId() != null) {
                    _payment_response.putString("apiKeyId", _response.getApiKeyId());
                }
                if (_response.getMerchantOrderId() != null) {
                    _payment_response.putString("merchantOrderId", _response.getMerchantOrderId());
                }
                if (_response.getCreationDate() != null) {
                    _payment_response.putString("creationDate", _response.getCreationDate());
                }
                if (_response.getTransactionId() != null) {
                    _payment_response.putString("transactionId", _response.getTransactionId());
                }
                if (_response.getProviderType() != null) {
                    _payment_response.putString("providerType", _response.getProviderType());
                }
                if (_response.getMerchantId() != null) {
                    _payment_response.putString("merchantId", _response.getMerchantId());
                }
                if (_response.getOrderReference() != null) {
                    _payment_response.putString("orderReference", _response.getOrderReference());
                }
                if (_response.getRefId() != null) {
                    _payment_response.putString("refId", _response.getRefId());
                }
                if (_response.getStatus() != null) {
                    _payment_response.putString("status", _response.getStatus());
                }

                if (_response.getError() != null) {
                    WritableMap _error = new WritableNativeMap();
                    if (_response.getError().getCause() != null) {
                        _error.putString("cause", _response.getError().getCause());
                    }
                    if (_response.getError().getExplanation() != null) {
                        _error.putString("explanation", _response.getError().getExplanation());
                    }
                    _payment_response.putMap("error", _error);
                }

                if (_response.getCard() != null) {
                    Card _responseCard = _response.getCard();
                    WritableMap _responseCardMap = new WritableNativeMap();


                    if (_responseCard.getResult() != null) {
                        _responseCardMap.putString("result", _responseCard.getResult());
                    }
                    if (_responseCard.getCurrency() != null) {
                        _responseCardMap.putString("currency", _responseCard.getCurrency());
                    }
                    _responseCardMap.putDouble("amount", _responseCard.getAmount());
                    if (_responseCard.getCardInfo() != null) {
                        WritableMap _cardInfoMap = new WritableNativeMap();
                        CardInfo _cardInfo = _responseCard.getCardInfo();
                        if (_cardInfo.getCardDataToken() != null) {
                            _cardInfoMap.putString("cardDataToken", _cardInfo.getCardDataToken());
                        }
                        if (_cardInfo.getCardHash() != null) {
                            _cardInfoMap.putString("cardHash", _cardInfo.getCardHash());
                        }
                        if (_cardInfo.getMaskedCard() != null) {
                            _cardInfoMap.putString("maskedCard", _cardInfo.getMaskedCard());
                        }
                        if (_cardInfo.getCardHolderName() != null) {
                            _cardInfoMap.putString("cardHolderName", _cardInfo.getCardHolderName());
                        }
                        if (_cardInfo.getExpiryMonth() != null) {
                            _cardInfoMap.putString("expiryMonth", _cardInfo.getExpiryMonth());
                        }
                        if (_cardInfo.getExpiryYear() != null) {
                            _cardInfoMap.putString("expiryYear", _cardInfo.getExpiryYear());
                        }
                        if (_cardInfo.getCardBrand() != null) {
                            _cardInfoMap.putString("cardBrand", _cardInfo.getCardBrand());
                        }
                        if (_cardInfo.getCcvToken() != null) {
                            _cardInfoMap.putString("ccvToken", _cardInfo.getCcvToken());
                        }
                        _responseCardMap.putMap("cardInfo", _cardInfoMap);
                    }
                    if (_responseCard.getMerchant() != null) {
                        WritableMap _merchantMap = new WritableNativeMap();
                        if (_responseCard.getMerchant().getMerchantRedirectUrl() != null) {
                            _merchantMap.putString("merchantRedirectUrl", _responseCard.getMerchant().getMerchantRedirectUrl());
                        }
                        _responseCardMap.putMap("merchant", _merchantMap);
                    }
                    if (_responseCard.getDSecure() != null) {
                        WritableMap _3dSecureMap = new WritableNativeMap();
                        DSecure _3DSecure = _responseCard.getDSecure();

                        if (_3DSecure.getAcsUrl() != null) {
                            _3dSecureMap.putString("acsUrl", _3DSecure.getAcsUrl());
                        }
                        if (_3DSecure.getProcessACSRedirectURL() != null) {
                            _3dSecureMap.putString("processACSRedirectURL", _3DSecure.getProcessACSRedirectURL());
                        }
                        if (_3DSecure.getRequestType() != null) {
                            _3dSecureMap.putString("requestType", _3DSecure.getRequestType());
                        }
                        if (_3DSecure.getPaReq() != null) {
                            _3dSecureMap.putString("paReq", _3DSecure.getPaReq());
                        }
                        _responseCardMap.putMap("DSecure", _3dSecureMap);
                    }

                    _payment_response.putMap("card", _responseCardMap);
                }

                _paymentResponseMap.putMap("response", _payment_response);
            }
        }
        return _paymentResponseMap;
    }
}

File: ios/KashierPaymentResponseParser.m
//
//  KashierPaymentResponseParser.m
//  react-native-kashier
//
//  Created by Meena on 7/16/20.
//

#import "KashierPaymentResponseParser.h"

@implementation  KashierPaymentResponseParser

+ (NSDictionary*) getMessagesFromResponse: (GenericResponse*) paymentResponse{
	NSMutableDictionary* parsedResponse = [[NSMutableDictionary alloc]init];
	P_Messages* _messages = NULL;

	if(paymentResponse !=NULL){
		if(paymentResponse.messages!=NULL){
			_messages = paymentResponse.messages;
		}
		if(paymentResponse.responseModel!=NULL){
			if(paymentResponse.responseModel.messages!=NULL){
				_messages = paymentResponse.responseModel.messages;
			}
			if(paymentResponse.responseModel.message!=NULL && paymentResponse.responseModel.message.messages!=NULL){
				_messages = paymentResponse.responseModel.message.messages;
			}
		}

	}
	if(_messages != NULL){
		if(_messages.ar!=NULL){
			[parsedResponse setObject: _messages.ar forKey:@"ar"];
		}
		if(_messages.en!=NULL){
			[parsedResponse setObject: _messages.en forKey:@"en"];
		}
	}
	return parsedResponse;
}

+ (NSString*) getResponseStatusFromString: (NSString*) str{
	NSString* status = @"UNKNOWN";
	if([str isEqualToString:@"SUCCESS"] ||
	   [str isEqualToString:@"200"]){
		status = @"SUCCESS";
	}else if([str isEqualToString:@"FAILURE"] ||
			 [str isEqualToString:@"INVALID_REQUEST"] ||
			 [str isEqualToString:@"PENDING"] ||
			 [str isEqualToString:@"PENDING_ACTION"]){
		status = str;
	}
	return status;
}

+ (NSString*) getResponseStatusFromResponse: (GenericResponse*) paymentResponse{
	NSString* _responseStatus = @"UNKNOWN";
	if(paymentResponse!=NULL){
		if(paymentResponse.status!=NULL && [KashierUtils isNonEmptyString:paymentResponse.status]){
			_responseStatus =  [self getResponseStatusFromString:paymentResponse.status];
		}
		if(paymentResponse.responseModel!=NULL){
			if(paymentResponse.responseModel.status!=NULL &&
			   [KashierUtils isNonEmptyString:paymentResponse.responseModel.status]){
				_responseStatus = [self getResponseStatusFromString:paymentResponse.responseModel.status];
			}
			if(paymentResponse.responseModel.message!=NULL && paymentResponse.responseModel.message.status!=NULL &&
			   [KashierUtils isNonEmptyString:paymentResponse.responseModel.message.status]){
				_responseStatus = [self getResponseStatusFromString:paymentResponse.responseModel.message.status];
			}
			if(paymentResponse.responseModel.cardModel!=NULL && paymentResponse.responseModel.cardModel.result!=NULL){
				_responseStatus = [self getResponseStatusFromString:paymentResponse.responseModel.cardModel.result];
			}
		}
	}
	return _responseStatus;
}

+ (NSDictionary*) parsePaymentResponse: (Payment*) paymentResponse{
	NSMutableDictionary* parsedResponse =[[NSMutableDictionary alloc]init];
	if(paymentResponse!=NULL){
		[parsedResponse setObject: [KashierPaymentResponseParser getMessagesFromResponse:paymentResponse] forKey:@"messages"];
		[parsedResponse setObject:[self getResponseStatusFromResponse:paymentResponse] forKey:@"status"];
		if(paymentResponse.response!=NULL){
			NSMutableDictionary* response =[[NSMutableDictionary alloc]init];
			if(paymentResponse.response.method!=NULL){
				[response setObject:paymentResponse.response.method forKey:@"method"];
			}
			if(paymentResponse.response.orderID!=NULL){
				[response setObject:paymentResponse.response.orderID forKey:@"orderId"];
			}
			if(paymentResponse.response.apiKeyID!=NULL){
				[response setObject:paymentResponse.response.apiKeyID forKey:@"apiKeyId"];
			}
			if(paymentResponse.response.merchantOrderID!=NULL){
				[response setObject:paymentResponse.response.merchantOrderID forKey:@"merchantOrderId"];
			}
			if(paymentResponse.response.creationDate!=NULL){
				[response setObject:paymentResponse.response.creationDate forKey:@"creationDate"];
			}
			if(paymentResponse.response.transactionID!=NULL){
				[response setObject:paymentResponse.response.transactionID forKey:@"transactionId"];
			}
			if(paymentResponse.response.providerType!=NULL){
				[response setObject:paymentResponse.response.providerType forKey:@"providerType"];
			}
			if(paymentResponse.response.merchantID!=NULL){
				[response setObject:paymentResponse.response.merchantID forKey:@"merchantId"];
			}
			if(paymentResponse.response.orderReference!=NULL){
				[response setObject:paymentResponse.response.orderReference forKey:@"orderReference"];
			}
			if(paymentResponse.response.refID!=NULL){
				[response setObject:paymentResponse.response.refID forKey:@"refId"];
			}
			if(paymentResponse.response.status!=NULL){
				[response setObject:paymentResponse.response.status forKey:@"status"];
			}
			if(paymentResponse.response.error!=NULL){
				NSMutableDictionary* _error = [[NSMutableDictionary alloc]init];
				if(paymentResponse.response.error.cause!=NULL &&
				   [KashierUtils isNonEmptyString:paymentResponse.response.error.cause]){
					[_error setObject:paymentResponse.response.error.cause forKey:@"cause"];
				}
				if(paymentResponse.response.error.explanation!=NULL &&
				   [KashierUtils isNonEmptyString:paymentResponse.response.error.explanation]){
					[_error setObject:paymentResponse.response.error.explanation forKey:@"explanation"];
				}
				[response setObject:_error forKey:@"error"];
			}

			if(paymentResponse.response.card!=NULL){
				ResponseCard* _card = paymentResponse.response.card;
				NSMutableDictionary* parsedCard = [[NSMutableDictionary alloc]init];
				if(_card.result!=NULL){
					[parsedCard setObject:_card.result forKey:@"result"];
				}
				if(_card.currency!=NULL){
					[parsedCard setObject:_card.currency forKey:@"currency"];
				}
				[parsedCard setObject:[[NSNumber numberWithDouble:_card.amount] stringValue] forKey:@"amount"];

				if(_card.cardInfo!=NULL){
					NSMutableDictionary* parsedCardInfo = [[NSMutableDictionary alloc]init];
					if(_card.cardInfo.cardDataToken !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.cardDataToken forKey:@"cardDataToken"];
					}
					if(_card.cardInfo.cardHash !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.cardHash forKey:@"cardHash"];
					}
					if(_card.cardInfo.maskedCard !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.maskedCard forKey:@"maskedCard"];
					}
					if(_card.cardInfo.cardHolderName !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.cardHolderName forKey:@"cardHolderName"];
					}
					if(_card.cardInfo.expiryMonth !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.expiryMonth forKey:@"expiryMonth"];
					}
					if(_card.cardInfo.expiryYear !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.expiryYear forKey:@"expiryYear"];
					}
					if(_card.cardInfo.cardBrand !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.cardBrand forKey:@"cardBrand"];
					}
					if(_card.cardInfo.ccvToken !=NULL){
						[parsedCardInfo setObject:_card.cardInfo.ccvToken forKey:@"ccvToken"];
					}

					[parsedCard setObject:parsedCardInfo forKey:@"cardInfo"];
				}

				if(_card.merchant!=NULL && _card.merchant.merchantRedirectURL!=NULL){
					NSMutableDictionary* _merchant = [[NSMutableDictionary alloc]init];
					[_merchant setObject:_card.merchant.merchantRedirectURL forKey:@"merchantRedirectUrl"];
					[parsedCard setObject:_merchant forKey:@"merchant"];
				}

				if(_card.card3DSecure!=NULL){
					NSMutableDictionary* _cardDSecure = [[NSMutableDictionary alloc]init];
					if(_card.card3DSecure.acsURL!=NULL){
						[_cardDSecure setObject:_card.card3DSecure.acsURL forKey:@"acsUrl"];
					}
					if(_card.card3DSecure.processACSRedirectURL!=NULL){
						[_cardDSecure setObject:_card.card3DSecure.processACSRedirectURL forKey:@"processACSRedirectURL"];
					}
					if(_card.card3DSecure.requestType!=NULL){
						[_cardDSecure setObject:_card.card3DSecure.requestType forKey:@"requestType"];
					}
					if(_card.card3DSecure.paReq!=NULL){
						[_cardDSecure setObject:_card.card3DSecure.paReq forKey:@"paReq"];
					}
					[parsedCard setObject:_cardDSecure forKey:@"DSecure"];
				}
				[response setObject:parsedCard forKey:@"card"];
			}
			[parsedResponse setObject:response forKey:@"response"];
		}
	}
	return parsedResponse;
}

@end
